src/lib.rs

```diff
@@ -1,8 +1,530 @@
-#[macro_use] extern crate log;
+#[cfg(feature = "cli")] use std::str::FromStr;
+
+use heck::ToUpperCamelCase;
+use k8s_openapi::apiextensions_apiserver::pkg::apis::apiextensions::v1::{
+    CustomResourceDefinition, CustomResourceDefinitionVersion,
+};
+use kube::{core::Version, ResourceExt};
 
 mod analyzer;
-pub use analyzer::{analyze, Config};
-mod output;
-pub use output::{format_docstr, Container, MapType, Member, Output};
+
 mod derive;
-pub use derive::Derive;
+mod output;
+
+pub use self::{
+    analyzer::{analyze, Config},
+    derive::Derive,
+    output::{format_docstr, Container, MapType, Member, Output},
+};
+
+/// Supported values for `kube`'s [`schema`] attribute.
+///
+/// [schema]: https://docs.rs/kube/latest/kube/derive.CustomResource.html#kubeschema--mode
+#[derive(
+    // std
+    Eq,
+    Copy,
+    Clone,
+    Debug,
+    Default,
+    PartialEq,
+    // strum
+    strum::Display,
+    strum::AsRefStr,
+)]
+#[cfg_attr(feature = "cli", derive(strum::EnumString))]
+#[strum(ascii_case_insensitive, serialize_all = "lowercase")]
+pub enum SchemaMode {
+    /// Instruct `kube` to expect `JsonSchema` to be implemented manually for `kopium` generated type(s).
+    Manual,
+
+    /// Instruct `kube` to automatically derive a `JsonSchema` implementation for `kopium` generated type(s).
+    Derived,
+
+    /// Instruct `kube` to skip deriving a `JsonSchema` implementation for `kopium` generated type(s) entirely.
+    ///
+    /// **NOTE**: the resulting CRD cannot be applied to a cluster without manual fiddling to add an OpenAPI schema.
+    #[default]
+    Disabled,
+}
+
+
+#[cfg_attr(feature = "cli", derive(clap::Args))]
+#[derive(Clone, Debug, typed_builder::TypedBuilder)]
+#[builder(
+    field_defaults(default),
+    mutators(
+        /// Add the supplied [`Derive`] directive to the list
+        /// of traits to be derived on generated types
+        pub fn derive(&mut self, value: Derive) {
+            self.derive_traits.push(value);
+        }
+
+        /// Add all the supplied [`Derive`] directives to the
+        /// list of traits to be derived  on generated types
+        pub fn derive_all(&mut self, values: impl IntoIterator<Item = Derive>) {
+            self.derive_traits.extend(values);
+        }
+    )
+)]
+pub struct TypeGenerator {
+    /// Use this CRD version if multiple versions are present
+    #[cfg_attr(feature = "cli", arg(long))]
+    pub api_version: Option<String>,
+
+    /// Do not emit prelude(s)
+    #[cfg_attr(feature = "cli", arg(long))]
+    pub hide_prelude: bool,
+
+    /// Do not derive CustomResource nor set kube-derive attributes
+    ///
+    /// If this is set, it makes any kube-derive specific options such as `--schema` unnecessary
+    #[cfg_attr(feature = "cli", arg(long))]
+    pub hide_kube: bool,
+
+    /// Emit doc comments from CRD field descriptions
+    #[cfg_attr(feature = "cli", arg(short = 'd', long = "docs"))]
```

Comment:
nice use of short/long to keep cli api and enriching the lib api

Reply:
<img width="350" src="https://c.tenor.com/33MP7vrlXJoAAAAC/tenor.gif"/>